### DISCUSSION TITLE
Release schedule

### DISCUSSION THREAD

**Discussion by vincentarelbundock** [easystats/insight#444] [General]
Sorry to bug you with this, but I'm curious to know if there's a plan for a next release in the near future (I see that they are quite frequent on the archive page).

I don't mean to put pressure at all. Just thinking about coordinating a release with `insight` dependency and trying to figure out which version I can require.

Again, no pressure! I know releasing to CRAN can be a pain.

**Comment by bwiernik** [comment]
Follow the "dataverse cycle" issue in easystats for where we are in the release cycles

**Comment by vincentarelbundock** [comment]
Got it, thanks!

https://github.com/easystats/easystats/issues/143

**Reply by strengejacke** [reply]
We have `easystats::on_CRAN()`, and I usually try to avoid submitting packages in red. That said, we could plan a new release soon:

![image](https://user-images.githubusercontent.com/26301769/136538413-87912522-3674-4e25-9930-01f78c5ecb7d.png)

What is your schedule for the next releases of your packages? I think it should be fine to coordinate these with _insight_, and submitting to CRAN is usually quite smooth, unless you have some breaking changes.

**Comment by vincentarelbundock** [comment]
Ah, the `on_cran` function is very useful, thanks!

I received an ominous email from Ripley saying that I need to release a new version before October 19th. 

(One of my tests fails at tolerance 0.000001 on the new ARM mac. Catastrophe!)

**Comment by vincentarelbundock** [comment]
@strengejacke I'm not sure what your process is for updating `NEWS.md`. But FYI, I skimmed through the commits since last release and added some stuff there.

**Reply by vincentarelbundock** [reply]
https://github.com/easystats/insight/commit/682236249ca9586f3186a53ff2027fc8b6452565

**Reply by bwiernik** [reply]
Thanks! Our process is "not as organized as it should be"

**Reply by vincentarelbundock** [reply]
I chuckled.

**Comment by strengejacke** [comment]
Any critical issues that need to be addressed before submitting to CRAN? Else, I might submit an _insight_ update this week.

**Reply by bwiernik** [reply]
Maybe the glmmTMB fixes Vincent was working on?

**Reply by vincentarelbundock** [reply]
My guess is that the `glmmTMB` fixes are going to require a little bit of infrastructure. I can take a look at that eventually, but my next couple weeks are going to be busy at work, so I'm unlikely to have as much time for hobby coding.

In any case, this only applies to a small class models, and we wouldn't be introducing *new* bugs. So I suggest we raise a warning for zero-inflated models when `predict="expectation"`, and we take a bit more time to figure out what the most generalizable/solid infrastructure is going to be for `glmmTMB` and similar cases.

[Disclosure: I am certainly biased, since CRAN is going to remove `marginaleffects` if I don't update it, and I made the mistake of developing against `insight` devel features, so current master is all sorts of broken against `0.14.4`.]

**Reply by vincentarelbundock** [reply]
Aside from `glmmTMB`, I don't see other major issues, and I have no immediate plans for new features.

**Reply by vincentarelbundock** [reply]
FYI, I added the proposed warning. I can be removed when we implement the zero-inflation correction: https://github.com/easystats/insight/commit/dca9e72ef426584073ee866e5c487d0115b12a78

**Comment by strengejacke** [comment]
> [Disclosure: I am certainly biased, since CRAN is going to remove marginaleffects if I don't update it, and I made the mistake of developing against insight devel features, so current master is all sorts of broken against 0.14.4.]

Great, you're now following @IndrajeetPatil footsteps. CRAN team already knows that every new submission of any easystats package will 99% sure break any of @IndrajeetPatil packages ;-)

**Comment by strengejacke** [comment]
Maybe we can add **https://github.com/easystats/modelbased/issues/151#issuecomment-939546873**. I think this should be easy, and the regex stuff to detect transformations can be copied from https://github.com/strengejacke/ggeffects/blob/master/R/utils_ggpredict.R

**Comment by strengejacke** [comment]
Does anybody know, if I have a string `"log(x+1)"`, how can I create a function that does `function(x) log(x + 1)`? Or when I have `"log(x+5)"`, I would like to return a function `function(x) log(x + 5)`.

**Reply by IndrajeetPatil** [reply]
Something like this?

``` r
arg <- quote(x + 1)
rlang::call2("log", arg)
#> log(x + 1)

arg <- quote(x + 5)
rlang::call2("log", arg)
#> log(x + 5)
```

<sup>Created on 2021-10-15 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

**Reply by bwiernik** [reply]
Maybe `str2lang()` ?

**Comment by strengejacke** [comment]
Nope.

``` r
fun <- function(x) str2lang("log(x+2)")

# is
fun(.3)
#> log(x + 2)

# should be
log(.3 + 2)
#> [1] 0.8329091

arg <- quote(x + 1)
fun <- function(x) rlang::call2("log", arg)
fun(.3)
#> log(x + 1)
```

<sup>Created on 2021-10-15 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

**Comment by IndrajeetPatil** [comment]
Here is a working example:
``` r
library(rlang)

foo <- function(x, a = 1) {
  args <- list(x = quote(x + a))
  eval(call2(.fn = log, !!!args))
}

foo(0.3)
#> [1] 0.2623643

log(0.3 + 1)
#> [1] 0.2623643

foo(0.3, 2)
#> [1] 0.8329091

log(0.3 + 2)
#> [1] 0.8329091
```

<sup>Created on 2021-10-15 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

**Reply by strengejacke** [reply]
ok, how do you make this function work when you have a string `"log(x+3")`? That's the only string you have, and from this string, you need to build a function `foo(x)`, that computes `log(x+3)` (`log` is always expected, but the `+<number>` may vary).

**Reply by strengejacke** [reply]
Maybe something like 

``` r
foo <- function(x, a) {
  args <- list(x = quote(x + a))
  do.call(log, args)
}

n <- 3
foo(iris$Sepal.Length, eval(n)) |> head()
#> [1] 2.091864 2.066863 2.041220 2.028148 2.079442 2.128232
```

<sup>Created on 2021-10-15 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

**Reply by strengejacke** [reply]
but not sure if this will work outside the specific context/environment, hm...

**Reply by strengejacke** [reply]
What I exactly want:

1. We have `lm(log(y+2) ~x, data = d)`
2. The function `get_transformation()` extracts the response transformation and extracts the string `"log(x+2)"`.
3. Now `get_transformation()` should return a function `foo(x)`, that internally does exact the same thing as defined in the above string (i.e. `log(x+2)`.

**Reply by IndrajeetPatil** [reply]
So, if I understand correctly, you want a function such that if I enter a string `"log(x+2)"` into it as an argument, it outputs a **function** of the form: `function(x) log(x + 2)`, and **not** a language type object (`log(x + 2)`)?

**Reply by strengejacke** [reply]
yes. Just that you don't enter a string, but it's rather extracted from the model object. See these new two functions:

https://github.com/easystats/insight/blob/master/R/find_transformation.R
https://github.com/easystats/insight/blob/master/R/get_transformation.R

**Reply by strengejacke** [reply]
For any `lm(log(y+ <number>) ~ x, data = d)`, `find_transformation()` returns `"log(x+<number>)"`. And `get_transformation()` should return a function that accepts one argument, and then applies the transformation that was used in the formula.

**Reply by strengejacke** [reply]
The idea came from here: https://github.com/easystats/modelbased/issues/151#issuecomment-939546873

**Comment by strengejacke** [comment]
@vincentarelbundock @bwiernik @IndrajeetPatil 

Ok, I'm running final checks now and will submit _insight_ today. I think we can add more features to `get_transformation()` later, the most important stuff should work for now.

**Reply by vincentarelbundock** [reply]
That's fantastic! Thanks so much!

**Comment by strengejacke** [comment]
@DominiqueMakowski @bwiernik 
See https://win-builder.r-project.org/incoming_pretest/insight_0.14.5_20211016_105703/reverseDependencies/summary.txt

Would modebased be ready for submission the next days in order to resolve this issue?

**Reply by strengejacke** [reply]
wait, might be an issue in _insight_, actually.

**Comment by vincentarelbundock** [comment]
> wait, might be an issue in _insight_, actually.

On the one hand, rev deps are kind of a pain, but on the other hand it's kind of neat that we can benefit from the test suites of downstream packages. https://github.com/easystats/insight/commit/3ae2bd2a9bbd29f38b80dee9d1a72bf77aefad75 is another example of a minor bug that I had introduced in prior commit and which I discovered because of `marginaleffects`' tests

**Comment by strengejacke** [comment]
on CRAN :-)

**Reply by vincentarelbundock** [reply]
> on CRAN :-)

Yessss
